madara/crates/client/settlement_client/src/starknet/event.rs

```diff
@@ -0,0 +1,396 @@
+
+                            let nonce = event.data.get(1).ok_or_else(|| {
+                                SettlementClientError::Starknet(StarknetClientError::EventProcessing {
+                                    message: "Missing nonce in event data".to_string(),
+                                    event_id: "MessageSent".to_string(),
+                                })
+                            })?;
+                            self.processed_events.insert(*nonce);
+
+                            let event_data = event
+                                .block_number
+                                .ok_or_else(|| {
+                                    SettlementClientError::Starknet(StarknetClientError::EventProcessing {
+                                        message: "Unable to get block number from event".to_string(),
+                                        event_id: "MessageSent".to_string(),
+                                    })
+                                })
+                                .map(|block_number| {
+                                    let selector = event.data.first().ok_or_else(|| {
+                                        SettlementClientError::Starknet(StarknetClientError::EventProcessing {
+                                            message: "Missing selector in event data".to_string(),
+                                            event_id: "MessageSent".to_string(),
+                                        })
+                                    })?;
+                                    let from = event.keys.get(2).ok_or_else(|| {
+                                        SettlementClientError::Starknet(StarknetClientError::EventProcessing {
+                                            message: "Missing from_address in event keys".to_string(),
+                                            event_id: "MessageSent".to_string(),
+                                        })
+                                    })?;
+                                    let to = event.keys.get(3).ok_or_else(|| {
+                                        SettlementClientError::Starknet(StarknetClientError::EventProcessing {
+                                            message: "Missing to_address in event keys".to_string(),
+                                            event_id: "MessageSent".to_string(),
+                                        })
+                                    })?;
+                                    let message_hash = event.keys.get(1).ok_or_else(|| {
+                                        SettlementClientError::Starknet(StarknetClientError::EventProcessing {
+                                            message: "Missing message_hash in event keys".to_string(),
+                                            event_id: "MessageSent".to_string(),
+                                        })
+                                    })?;
+
+                                    Ok(L1toL2MessagingEventData {
```

Comment:
Same as earlier, adding a `TryFrom` or other conversion utility fn would help keep `poll_next()` easier to audit/maintain